src/lib/scheduler/node_queue_scheduler.hpp

```diff
@@ -12,63 +12,57 @@ namespace hyrise {
 /*
  * GENERAL SCHEDULING CONCEPT
  *
- * Everything that needs to be processed is encapsulated in tasks. For example, in the context of the database
- * the OperatorTasks encapsulates database operators (here, it only encapsulates the execute function).
- * A task will then be pushed by a Scheduler into a TaskQueue and pulled out by a Worker to be processed.
+ * The scheduler is the main entry point and there are currently two alternative implementations: the
+ * ImmediateExecutionScheduler (single-threaded, for benchmarking and debugging) and the NodeQueueScheduler (multi-
+ * threaded).
+ *
+ * Everything that needs to be processed is encapsulated in tasks. A task will be pushed into a TaskQueue by a
+ * scheduler and pulled by a Worker to be processed.
  *
  *
  * TASK DEPENDENCIES
  *
- * Tasks can be dependent of each other. For example, in the context of the database, a table scan operation can be
- * dependent on a GetTable operation and so do the tasks that encapsulates these operations.
- * Therefore, not every task that is on top of the TaskQueue is ready to be processed. Pulling this unready task and
- * pushing it back to the queue could result in high latency for this task and in the context of databases in a
- * high latency for queries. Not pulling this unready task and waiting until it is ready would block the queue.
- * To avoid these situations, a TaskQueue is implemented as a vector and will be iterated until a ready task is found.
- * This preserves the desired order and will not block the queue.
- * To determine if a task is ready, it checks its parent tasks if they are done. A task will be set done after it
- * was processed successfully.
+ * Tasks can be dependent on each other. For example, a TableScan operation can depend on a GetTable operation and so
+ * do the tasks that encapsulate these operations. Tasks with predecessors are not scheduled (i.e., not added to the
```

Comment:
`Tasks with predecessors are not scheduled ...` – can we make this a bit clearer? What does this mean for the tasks of a query?

Reply:
Better?

```
 * Tasks can be dependent on each other. For example, a TableScan operation can depend on a GetTable operation and so
 * do the tasks that encapsulate these operations. Tasks with predecessors are not scheduled (i.e., not added to the
 * TaskQueues) to avoid workers from pulling tasks that cannot (yet) be processed. Instead, succeeding tasks are
 * executed as soon as the predecessing task has been processed (workers try to process all successors before pulling
 * new tasks from the TaskQueues).
```

Reply:
From reading it, this means we only schedule the leaf (!!!1einself) tasks of a query plan and execute all subsequent tasks directly if possible, right? So in general, this tiny bit of information tells us that we, in fact, schedule rather a whole query than a single operator, doesn't it?